File: sdks/typescript/src/core/signed-payload.ts
/**
 * DCP v2.0 SignedPayload envelope.
 *
 * The artifact (payload) never contains signature data. Signatures live in a
 * sibling field. Canonicalization is always over `payload` — no field
 * stripping needed.
 */

import type { CompositeSignature } from './composite-sig.js';
import { canonicalizeV2 } from './canonicalize.js';
import { sha256Hex } from './dual-hash.js';

export interface SignedPayload<T = unknown> {
  payload: T;
  payload_hash: string;
  composite_sig: CompositeSignature;
}

/**
 * Compute the canonical bytes and hash for a payload.
 * Returns { canonicalBytes, payloadHash } where payloadHash is "sha256:<hex>".
 */
export function preparePayload(payload: unknown): {
  canonicalBytes: Uint8Array;
  payloadHash: string;
} {
  const canonical = canonicalizeV2(payload);
  const bytes = new TextEncoder().encode(canonical);
  const hash = sha256Hex(bytes);
  return {
    canonicalBytes: bytes,
    payloadHash: `sha256:${hash}`,
  };
}

/**
 * Verify that a SignedPayload's payload_hash matches the actual payload.
 */
export function verifyPayloadHash(signed: SignedPayload): boolean {
  const { payloadHash } = preparePayload(signed.payload);
  return signed.payload_hash === payloadHash;
}

File: sdks/python/dcp_ai/v2/signed_payload.py
"""
Signed payload preparation and verification for DCP v2.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_sig import CompositeSignature
from dcp_ai.v2.dual_hash import sha256_hex


@dataclass
class SignedPayloadData:
    payload: Any
    payload_hash: str
    composite_sig: CompositeSignature


def prepare_payload(payload: Any) -> tuple[bytes, str]:
    """Canonicalize a payload and compute its prefixed hash.

    Returns (canonical_bytes, payload_hash) where payload_hash is "sha256:<hex>".
    """
    canonical = canonicalize_v2(payload)
    canonical_bytes = canonical.encode("utf-8")
    hex_digest = sha256_hex(canonical_bytes)
    return canonical_bytes, f"sha256:{hex_digest}"


def verify_payload_hash(signed: dict[str, Any]) -> bool:
    """Re-canonicalize signed["payload"] and check it matches signed["payload_hash"]."""
    _, expected_hash = prepare_payload(signed["payload"])
    return signed.get("payload_hash") == expected_hash
